internal/jobs/launch_instance_aws.go

```diff
@@ -0,0 +1,91 @@
+package jobs
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/RHEnVision/provisioning-backend/internal/clients/ec2"
+	"github.com/RHEnVision/provisioning-backend/internal/clients/sts"
+	"github.com/RHEnVision/provisioning-backend/internal/ctxval"
+	"github.com/RHEnVision/provisioning-backend/internal/dao"
+	"github.com/aws/aws-sdk-go-v2/service/ec2/types"
+	"github.com/lzap/dejq"
+)
+
+type LaunchInstanceAWSTaskArgs struct {
+	AccountID     int64  `json:"account_id"`
+	PubkeyID      int64  `json:"pubkey_id"`
+	ReservationID int64  `json:"reservation_id"`
+	ImageID       string `json:"image_id"`
+	Amount        int32  `json:"amount"`
+	InstanceType  string `json:"instance_type"`
+	ARN           string `json:"arn"`
+	PubkeyHandle  string `json:"pubkey_handle"`
+}
+
+func EnqueueLaunchInstanceAWS(ctx context.Context, args *LaunchInstanceAWSTaskArgs) error {
+	logger := ctxval.Logger(ctx)
+	logger.Debug().Interface("arg", args).Msgf("Enqueuing launch instance AWS job: %+v", args)
+
+	pj := dejq.PendingJob{
+		Type: TypeLaunchInstanceAws,
+		Body: args,
+	}
+	err := Queue.Enqueue(ctx, pj)
+	if err != nil {
+		return fmt.Errorf("unable to enqueue: %w", err)
+	}
+
+	return nil
+}
+
+func HandleLaunchInstanceAWS(ctx context.Context, job dejq.Job) error {
+	ctxLogger := ctxval.Logger(ctx)
+	ctxLogger.Debug().Msg("Started launch instance AWS job")
+
+	args := LaunchInstanceAWSTaskArgs{}
+	err := job.Decode(&args)
+	if err != nil {
+		ctxLogger.Error().Err(err).Msg("unable to decode arguments")
+		return fmt.Errorf("unable to decode args: %w", err)
+	}
+
+	ctx = ctxval.WithAccountId(ctx, args.AccountID)
+	logger := ctxLogger.With().Int64("reservation", args.ReservationID).Logger()
+	logger.Info().Interface("args", args).Msg("Processing launch instance AWS job")
+
+	// upload to cloud with a tag
+	client := ec2.NewEC2Client(ctx)
+	stsClient, err := sts.NewSTSClient(ctx)
+	if err != nil {
+		return fmt.Errorf("cannot initialize sts client: %w", err)
+	}
+
+	crd, err := stsClient.AssumeRole(args.ARN)
+	if err != nil {
+		return fmt.Errorf("cannot assume role: %w", err)
+	}
+
+	newEC2Client, err := client.CreateEC2ClientFromConfig(crd)
+	if err != nil {
+		return fmt.Errorf("cannot create new ec2 client from config: %w", err)
+	} //get the info needed to launch and instance and pass it
+	//TODO: NEEDS THE RESOURCE NAME LIKE "lzap-rsa-2010-{{$randomInt}}"
+	// Should be added to args.PubkeyHandle, the name is not saved?
+	_, err = newEC2Client.RunInstances(ctx, args.Amount, types.InstanceType(args.InstanceType), args.ImageID, "lzap-rsa-2010-{{$randomInt}}")
+	if err != nil {
+		return fmt.Errorf("cannot run instances: %w", err)
+	}
+
+	// mark the reservation as finished
```

Comment:
So, this is the tricky part. In order to mark the reservation as finished, we need to confirm the instance has been launched. I haven't looked into RunInstances response yet, so let's do it now.

https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_RunInstances.html

Looks like the call returns a reservation id, which is perhaps a good handle to store in the AWS reservation detail table for debugging purposes (in case the customer needs to open up a case for AWS Support).

Then it returns an array of instances: https://docs.aws.amazon.com/AWSEC2/latest/APIReference/API_Instance.html

This means we cannot store a single instance id in AWS reservation detail, we need a table relationship for these instance IDs.